lib/Transforms/Instrumentation/ExtendedPointer.cc

```diff
@@ -201,41 +206,70 @@ bool ExtendedPointer::instrument(Value *Ptr, Value *InstVal,
 /* Record information of address Ptr, store/update the base address and size */
 bool ExtendedPointer::instrumentAddress(Value *Ptr, const DataLayout &DL,
                                         Value *BasePtr) {
+  Ptr->setName("raw_ptr");
+  Type *resultType;
   if (!BasePtr) {
-    Type *IntTy = DL.getIntPtrType(Ptr->getType());
-    // set_fat_slot0(Ptr, Base)
-    Builder->CreateCall(
+    if (auto *ALI = dyn_cast<AllocaInst>(Ptr)) {
+       resultType = ALI->getAllocatedType();
+    } else {
+      llvm_unreachable("only handling GEP instructions");
+    }
+    CallInst* withBase = Builder->CreateCall(
       m_setFatSlot0,
       {
-        Builder->CreateBitCast(Ptr, Builder->getInt8PtrTy()),
-        Builder->CreatePtrToInt(Ptr, Builder->getInt32Ty())
+        Constant::getNullValue(Builder->getInt8PtrTy()),
+        ConstantInt::get(IntptrTy, 0)
+      }
+    );
+    // set_fat_slot1(Ptr, Size)
+    CallInst* withSize = Builder->CreateCall(
+      m_setFatSlot1,
+      {
+        Constant::getNullValue(Builder->getInt8PtrTy()),
+        ConstantInt::get(IntptrTy, 0)
       }
     );
+    Value* casted = Builder->CreateBitCast(withSize, resultType->getPointerTo());
+    Ptr->replaceAllUsesWith(casted);
     Value *sizeVal;
     // alloca
     if (auto *ALI = dyn_cast<AllocaInst>(Ptr)) {
       uint64_t size = DL.getTypeStoreSize(ALI->getAllocatedType());
-      sizeVal = ConstantInt::get(Type::getInt32Ty(*Ctx), size);
+      sizeVal = ConstantInt::get(IntptrTy, size);
     } else {
       llvm_unreachable("unexpected lack of base address");
     }
-    // set_fat_slot1(Ptr, Size)
-    Builder->CreateCall(
-      m_setFatSlot1,
-      {
-        Builder->CreateBitCast(Ptr, Builder->getInt8PtrTy()),
-        sizeVal
-      }
-    );
+    // Type *IntTy = DL.getIntPtrType(Ptr->getType());
+    // set_fat_slot0(Ptr, Base)
+    Builder->SetInsertPoint(withBase);
```

Comment:
lost at what this code is trying to do. need a comment

Reply:
This is also leftover design when I was working on the fat ptr encoding implementation. Needed to replace all usage of `Ptr` with ones with base and size info encoded  (`withSize`); I can revert this pattern if we are just planning to use hashmap implementation.